specs/01-sharded-archiving.md

```diff
@@ -0,0 +1,288 @@
+  the block from the child shard to the history of the parent by including it on-chain into the
+  `shard_blocks[]` table.
+- To verify the `BlockInfo`, nodes in the parent shard:
+  - Ensure `block_index` is the immediate block after last committed block number in the parent
+    shard.
+- Output: `BlockInfo` from child shard included in `shard_blocks[]` table of the parent shard.
+
+#### 6. Segment Construction
+
+- Input: Buffered data of size `RAW_RECORD_SIZE`.
+- When the buffer (after encoding) contains enough data to fill a record of `RAW_RECORD_SIZE` bytes,
+  it is archived:
+  1. Split the record into `d = 2^15 =NUM_CHUNKS` chunks of size `SAFE_BYTES`.
+  2. Perform the same operations to extend with parity chunks and commit to the chunks as was done
+     for the block. This is done by creating a Merkle tree of the chunks of a record, and committing
+     to it.
+- Erasure code records column-wise with `extend(column, ERASURE_CODING_RATE)`. This effectively
+  doubles the number of rows and thus, records per segment to `NUM_PIECES` (256).
+- For each row, a proof of the row chunks is created by building a Merkle Proof of the individual
+  chunks in the share, `Cs_i`, along with their proof `ws_i`.
+- For each record, form a `piece = record || record_proof || record_proof`, `(ri∣∣Ci∣∣πi)`
+- Compute the `segment_root` by computing the Merkle proof of all the record proofs of all rows, the
+  raw records and the extended records `C_segment = Merkle_Proof(C0, C1, ..., Cn, C00, Cnn)`.
+- Append the `segment_root` to the global `segment_proofs[]` table of the chain.
+- The segment now consists of `NUM_PIECES` records of roughly 1MiB each
+  (`32 bytes * 2^15 chunks + 32 bytes proof + ~120 bytes proof`), `NUM_PIECES` piece
+  proofs, `NUM_PIECES` proofs of 32 bytes each and one 32-byte `segment_root`.
+- New pieces need to be stored by farmers and made available for retrieval by other shards and nodes
+  in the system.
+- Pieces from a shard are uniquely identified through an sequentially increasing `piece_index`.
+- Output: Segment with `NUM_PIECES` pieces and a `segment_root`.
+
+#### 7. Segment proof Submission to Beacon Chain
+
+- Input: New segments from a shard.
+- As soon as a new block is produced that includes the creation of a new segment, the segment proof
+  is submitted as a transaction in the beacon chain to commit the segment into the global history of
+  the system.
+- To submit this information to the beacon chain a `SegmentInfo` transaction is created that
+  includes the `segment_root`, the `segment_index`, and the `shard_id` of the shard where the
+  segment was created.
+- Any node in the shard (from farmers to full-nodes) are allowed to create the `SegmentInfo`
+  transaction that submits the segments into the parent's shard history.
+
+```rust
+type SegmentInfo = {
+  segment_root: Vec<u8>, // 32B
+  segment_index: u64, // 8B
+  shard_id: ShardId, // 4B
+}
```

Comment:
```suggestion
struct SegmentInfo {
  shard_id: ShardId, // 4B
  segment_index: u64,
  segment_root: [u8; 32],
}
```

Also segment index is currently a global thing, so it should probably be called differently or I'm not sure what you meant there, but child shard doesn't know the index until it is included in the beacon chain.

Reply:
I was assuming that segments will also keep a local index to determine their order within the shard, even if once they are committed they are assigned a global index.

Reply:
Yes, that might be useful for syncing a specific shard indeed, but I'm not yet sure it is actually required for pushing up the stack